loki/transformations/block_index_transformations.py

```diff
@@ -0,0 +1,416 @@
+            elif suff == 'LM':
+                basetype = BasicType.LOGICAL
+            else:
+                basetype = BasicType.REAL
+
+            pointer_type = SymbolAttributes(basetype, pointer=True, kind=kind, shape=view_shape)
+            contig_pointer_type = pointer_type.clone(contiguous=True, shape=array_shape)
+
+            pointer_var = Variable(name='P', type=pointer_type, dimensions=view_shape)
+            contig_pointer_var = pointer_var.clone(name='P_FIELD', type=contig_pointer_type, dimensions=array_shape) # pylint: disable=no-member
+
+            decls = (ir.VariableDeclaration(symbols=(pointer_var,)),)
+            decls += (ir.VariableDeclaration(symbols=(contig_pointer_var,)),)
+
+            typedefs += (ir.TypeDef(name=_type, body=decls, parent=field_array_module),) # pylint: disable=unexpected-keyword-arg
+
+        return typedefs
+
+    def _create_dummy_field_api_defs(self, field_array_mod_imports):
+        """
+        Create dummy definitions for FIELD_API wrapper-types to enrich typedefs.
+        """
+
+        wrapper_types = {sym.name for imp in field_array_mod_imports for sym in imp.symbols}
+
+        # create dummy module with empty spec
+        field_array_module = Module(name='FIELD_ARRAY_MODULE', spec=ir.Section(body=()))
+
+        # build parkind1 import
+        parkind_import = self._build_parkind_import(field_array_module, wrapper_types)
+        field_array_module.spec.append(parkind_import)
+
+        # build dummy type definitions
+        typedefs = self._build_field_array_types(field_array_module, wrapper_types)
+        field_array_module.spec.append(typedefs)
+
+        return [field_array_module,]
+
+    @staticmethod
+    def propagate_defs_to_children(key, definitions, successors):
+        """
+        Enrich all successors with the dummy FIELD_API definitions.
+        """
+
+        for child in successors:
+            child.ir.enrich(definitions)
+            child.trafo_data.update({key: {'definitions': definitions}})
+
+    def process_driver(self, routine, successors):
+
+        # create dummy definitions for field_api wrapper types
+        field_array_mod_imports = [imp for imp in routine.imports if imp.module.lower() == 'field_array_module']
+        definitions = []
+        if field_array_mod_imports:
+            definitions += self._create_dummy_field_api_defs(field_array_mod_imports)
+
+        # propagate dummy field_api wrapper definitions to children
+        self.propagate_defs_to_children(self._key, definitions, successors)
+
+        #TODO: we also need to process any code inside a loki/acdc parallel pragma at the driver layer
+
+    def build_ydvars_global_gfl_ptr(self, var):
+        """Replace accesses to thread-local ``YDVARS%GFL_PTR`` with global ``YDVARS%GFL_PTR_G``."""
+
+        if (parent := var.parent):
+            parent = self.build_ydvars_global_gfl_ptr(parent)
+
+        _type = var.type
+        if 'gfl_ptr' in var.basename.lower():
```

Comment:
[no action] This is extremely bespoke to (potentially changing) IFS conventions, which is absolutely fine for now. In the future we will likely need to factor out bespoke "IFS Fields" transformations into a separate sub-package, as we add more to these.